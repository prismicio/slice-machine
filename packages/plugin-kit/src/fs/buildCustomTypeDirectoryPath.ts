import * as path from "node:path";

import {
	buildCustomTypeLibraryDirectoryPath,
	BuildCustomTypeLibraryDirectoryPathArgs,
} from "./buildCustomTypeLibraryDirectoryPath";

export type BuildCustomTypeDirectoryPathArgs = {
	customTypeID: string;
} & BuildCustomTypeLibraryDirectoryPathArgs;

export const buildCustomTypeDirectoryPath = (
	args: BuildCustomTypeDirectoryPathArgs,
): string => {
	return path.join(
		buildCustomTypeLibraryDirectoryPath(args),
		args.customTypeID,
	);
};
