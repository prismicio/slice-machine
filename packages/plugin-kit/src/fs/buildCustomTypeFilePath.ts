import * as path from "node:path";

import {
	buildCustomTypeDirectoryPath,
	BuildCustomTypeDirectoryPathArgs,
} from "./buildCustomTypeDirectoryPath";

export type BuildCustomTypeFilePathArgs = {
	filename: string;
} & BuildCustomTypeDirectoryPathArgs;

export const buildCustomTypeFilePath = (
	args: BuildCustomTypeFilePathArgs,
): string => {
	return path.join(buildCustomTypeDirectoryPath(args), args.filename);
};
