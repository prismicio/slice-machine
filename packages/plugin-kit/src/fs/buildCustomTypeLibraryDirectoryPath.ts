import { SliceMachineHelpers } from "../createSliceMachineHelpers";

import { CUSTOM_TYPES_LIBRARY_DIRNAME } from "./constants";

export type BuildCustomTypeLibraryDirectoryPathArgs = {
	absolute?: boolean;
	helpers: SliceMachineHelpers;
};

export const buildCustomTypeLibraryDirectoryPath = (
	args: BuildCustomTypeLibraryDirectoryPathArgs,
): string => {
	return args.absolute
		? args.helpers.joinPathFromRoot(CUSTOM_TYPES_LIBRARY_DIRNAME)
		: CUSTOM_TYPES_LIBRARY_DIRNAME;
};
