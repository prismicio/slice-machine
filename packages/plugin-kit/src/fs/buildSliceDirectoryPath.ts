import * as path from "node:path";

import { pascalCase } from "./lib/pascalCase";
import {
	resolveSliceModel,
	ResolveSliceModelArgs,
} from "./lib/resolveSliceModel";

import {
	buildSliceLibraryDirectoryPath,
	BuildSliceLibraryDirectoryPathArgs,
} from "./buildSliceLibraryDirectoryPath";

export type BuildSliceDirectoryPathArgs = BuildSliceLibraryDirectoryPathArgs &
	ResolveSliceModelArgs;

export const buildSliceDirectoryPath = async (
	args: BuildSliceDirectoryPathArgs,
): Promise<string> => {
	const model = await resolveSliceModel(args);

	return path.join(
		buildSliceLibraryDirectoryPath(args),
		pascalCase(model.name),
	);
};
