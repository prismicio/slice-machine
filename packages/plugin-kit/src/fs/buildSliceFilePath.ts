import * as path from "node:path";

import {
	buildSliceDirectoryPath,
	BuildSliceDirectoryPathArgs,
} from "./buildSliceDirectoryPath";

export type BuildSliceFilePathArgs = {
	filename: string;
} & BuildSliceDirectoryPathArgs;

export const buildSliceFilePath = async (
	args: BuildSliceFilePathArgs,
): Promise<string> => {
	return path.join(await buildSliceDirectoryPath(args), args.filename);
};
