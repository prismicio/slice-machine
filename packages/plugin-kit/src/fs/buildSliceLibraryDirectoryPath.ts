import * as path from "node:path";

import { SliceMachineHelpers } from "../createSliceMachineHelpers";

export type BuildSliceLibraryDirectoryPathArgs = {
	libraryID: string;
	absolute?: boolean;
	helpers: SliceMachineHelpers;
};

export const buildSliceLibraryDirectoryPath = (
	args: BuildSliceLibraryDirectoryPathArgs,
): string => {
	return args.absolute
		? args.helpers.joinPathFromRoot(args.libraryID)
		: path.join(args.libraryID);
};
