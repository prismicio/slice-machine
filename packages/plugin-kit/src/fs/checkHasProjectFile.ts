import { checkPathExists } from "./lib/checkPathExists";

import { SliceMachineHelpers } from "../createSliceMachineHelpers";

export type CheckHasProjectFileArgs = {
	filename: string;
	helpers: SliceMachineHelpers;
};

export const checkHasProjectFile = async (
	args: CheckHasProjectFileArgs,
): Promise<boolean> => {
	const filePath = args.helpers.joinPathFromRoot(args.filename);

	return checkPathExists(filePath);
};
