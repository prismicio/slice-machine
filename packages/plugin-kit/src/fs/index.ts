///////////////////////////////////////////////////////////////////////////////
//
// Project helpers
//
///////////////////////////////////////////////////////////////////////////////

export { checkIsTypeScriptProject } from "./checkIsTypeScriptProject";

export { upsertGlobalTypeScriptTypes } from "./upsertGlobalTypeScriptTypes";
export type { UpsertGlobalTypeScriptTypesArgs } from "./upsertGlobalTypeScriptTypes";

export { checkHasProjectFile } from "./checkHasProjectFile";
export type { CheckHasProjectFileArgs } from "./checkHasProjectFile";

export { writeProjectFile } from "./writeProjectFile";
export type { WriteProjectFileArgs } from "./writeProjectFile";

export { readProjectFile } from "./readProjectFile";
export type { ReadProjectFileArgs } from "./readProjectFile";

export { deleteProjectFile } from "./deleteProjectFile";
export type { DeleteProjectFileArgs } from "./deleteProjectFile";

export { readProjectEnvironment } from "./readProjectEnvironment";
export type {
	ReadProjectEnvironmentArgs,
	ReadEnvironmentReturnType,
} from "./readProjectEnvironment";

export { writeProjectEnvironment } from "./writeProjectEnvironment";
export type { WriteProjectEnvironmentArgs } from "./writeProjectEnvironment";

///////////////////////////////////////////////////////////////////////////////
//
// Slice helpers
//
///////////////////////////////////////////////////////////////////////////////

export { buildSliceDirectoryPath } from "./buildSliceDirectoryPath";
export type { BuildSliceDirectoryPathArgs } from "./buildSliceDirectoryPath";

export { buildSliceFilePath } from "./buildSliceFilePath";
export type { BuildSliceFilePathArgs } from "./buildSliceFilePath";

export { buildSliceLibraryDirectoryPath } from "./buildSliceLibraryDirectoryPath";
export type { BuildSliceLibraryDirectoryPathArgs } from "./buildSliceLibraryDirectoryPath";

export { readSliceLibrary } from "./readSliceLibrary";
export type {
	ReadSliceLibraryArgs,
	ReadSliceLibraryReturnType,
} from "./readSliceLibrary";

export { writeSliceFile } from "./writeSliceFile";
export type { WriteSliceFileArgs } from "./writeSliceFile";

export { writeSliceModel } from "./writeSliceModel";
export type { WriteSliceModelArgs } from "./writeSliceModel";

export { readSliceFile } from "./readSliceFile";
export type { ReadSliceFileArgs } from "./readSliceFile";

export { readSliceModel } from "./readSliceModel";
export type {
	ReadSliceModelArgs,
	ReadSliceModelReturnType,
} from "./readSliceModel";

export { renameSlice } from "./renameSlice";
export type { RenameSliceArgs } from "./renameSlice";

export { deleteSliceFile } from "./deleteSliceFile";
export type { DeleteSliceFileArgs } from "./deleteSliceFile";

export { deleteSliceDirectory } from "./deleteSliceDirectory";
export type { DeleteSliceDirectoryArgs } from "./deleteSliceDirectory";

///////////////////////////////////////////////////////////////////////////////
//
// Custom type helpers
//
///////////////////////////////////////////////////////////////////////////////

export { buildCustomTypeDirectoryPath } from "./buildCustomTypeDirectoryPath";
export type { BuildCustomTypeDirectoryPathArgs } from "./buildCustomTypeDirectoryPath";

export { buildCustomTypeFilePath } from "./buildCustomTypeFilePath";
export type { BuildCustomTypeFilePathArgs } from "./buildCustomTypeFilePath";

export { buildCustomTypeLibraryDirectoryPath } from "./buildCustomTypeLibraryDirectoryPath";
export type { BuildCustomTypeLibraryDirectoryPathArgs } from "./buildCustomTypeLibraryDirectoryPath";

export { readCustomTypeLibrary } from "./readCustomTypeLibrary";
export type {
	ReadCustomTypeLibraryArgs,
	ReadCustomTypeLibraryReturnType,
} from "./readCustomTypeLibrary";

export { writeCustomTypeFile } from "./writeCustomTypeFile";
export type { WriteCustomTypeFileArgs } from "./writeCustomTypeFile";

export { writeCustomTypeModel } from "./writeCustomTypeModel";
export type { WriteCustomTypeModelArgs } from "./writeCustomTypeModel";

export { readCustomTypeFile } from "./readCustomTypeFile";
export type { ReadCustomTypeFileArgs } from "./readCustomTypeFile";

export { readCustomTypeModel } from "./readCustomTypeModel";
export type {
	ReadCustomTypeModelArgs,
	ReadCustomTypeModelReturnType,
} from "./readCustomTypeModel";

export { renameCustomType } from "./renameCustomType";
export type { RenameCustomTypeArgs } from "./renameCustomType";

export { deleteCustomTypeFile } from "./deleteCustomTypeFile";
export type { DeleteCustomTypeFileArgs } from "./deleteCustomTypeFile";

export { deleteCustomTypeDirectory } from "./deleteCustomTypeDirectory";
export type { DeleteCustomTypeDirectoryArgs } from "./deleteCustomTypeDirectory";

export { readSliceTemplateLibrary } from "./readSliceTemplateLibrary";
export type { ReadSliceTemplateLibraryArgs } from "./readSliceTemplateLibrary";
