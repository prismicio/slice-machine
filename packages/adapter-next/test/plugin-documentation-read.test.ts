import { describe, expect, it } from "vitest";
import { createMockFactory } from "@prismicio/mock";
import * as fs from "node:fs/promises";
import * as path from "node:path";

/**
 * !!! DO NOT use this mock factory in tests !!!
 *
 * @remarks
 * Its seed is not specific to be used outside the most general cases.
 */
const mock = createMockFactory({ seed: import.meta.url });

// Slice model to be used in general tests.
const model = mock.model.customType({
	id: "foo_bar",
	repeatable: true,
});

describe("PageSnippet", () => {
	describe("App Router", () => {
		it("return a snippet with TypeScript for TypeScript projects", async (ctx) => {
			await fs.writeFile(
				path.join(ctx.project.root, "tsconfig.json"),
				JSON.stringify({}),
			);

			const res = await ctx.pluginRunner.callHook("documentation:read", {
				kind: "PageSnippet",
				data: {
					model,
				},
			});

			const item = res.data.flat().find((item) => item.label === "App Router");

			expect(item).toMatchSnapshot();
		});

		it("return a snippet with JavaScript", async (ctx) => {
			const res = await ctx.pluginRunner.callHook("documentation:read", {
				kind: "PageSnippet",
				data: {
					model,
				},
			});

			const item = res.data.flat().find((item) => item.label === "App Router");

			expect(item).toMatchSnapshot();
		});
	});

	describe("Pages Router", () => {
		it("return a snippet with TypeScript for TypeScript projects", async (ctx) => {
			await fs.writeFile(
				path.join(ctx.project.root, "tsconfig.json"),
				JSON.stringify({}),
			);

			const res = await ctx.pluginRunner.callHook("documentation:read", {
				kind: "PageSnippet",
				data: {
					model,
				},
			});

			const item = res.data
				.flat()
				.find((item) => item.label === "Pages Router");

			expect(item).toMatchSnapshot();
		});

		it("return a snippet with JavaScript", async (ctx) => {
			const res = await ctx.pluginRunner.callHook("documentation:read", {
				kind: "PageSnippet",
				data: {
					model,
				},
			});

			const item = res.data
				.flat()
				.find((item) => item.label === "Pages Router");

			expect(item).toMatchSnapshot();
		});
	});
});
