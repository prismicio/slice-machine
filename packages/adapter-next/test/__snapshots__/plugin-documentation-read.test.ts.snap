// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PageSnippet > App Router > return a snippet with JavaScript 1`] = `
{
  "content": "## Create your Foo Bar's page component

Add a new route by creating an \`app/[uid]/page.js\` file. (If the route should be nested in a child directory, you can create the file in a directory, like \`app/marketing/[uid]/page.js\`.)

Paste in this code:

~~~js [app/[uid]/page.js]
import { notFound } from \\"next/navigation\\";
import { SliceZone } from \\"@prismicio/react\\";

import { createClient } from \\"@/prismicio\\";
import { components } from \\"@/slices\\";

export default async function Page({ params }) {
  const client = createClient();
  const page = await client
    .getByUID(\\"foo_bar\\", params.uid)
    .catch(() => notFound());

  return <SliceZone slices={page.data.slices} components={components} />;
}

export async function generateMetadata({ params }) {
  const client = createClient();
  const page = await client
    .getByUID(\\"foo_bar\\", params.uid)
    .catch(() => notFound());

  return {
    title: page.data.meta_title,
    description: page.data.meta_description,
  };
}

export async function generateStaticParams() {
  const client = createClient();
  const pages = await client.getAllByType(\\"foo_bar\\");

  return pages.map((page) => {
    return { uid: page.uid };
  });
}
~~~

Make sure all of your import paths are correct. See the [install guide](https://prismic.io/docs/setup-nextjs) for more information.",
  "label": "App Router",
}
`;

exports[`PageSnippet > App Router > return a snippet with TypeScript for TypeScript projects 1`] = `
{
  "content": "## Create your Foo Bar's page component

Add a new route by creating an \`app/[uid]/page.tsx\` file. (If the route should be nested in a child directory, you can create the file in a directory, like \`app/marketing/[uid]/page.tsx\`.)

Paste in this code:

~~~tsx [app/[uid]/page.tsx]
import { Metadata } from \\"next\\";
import { notFound } from \\"next/navigation\\";
import { SliceZone } from \\"@prismicio/react\\";

import { createClient } from \\"@/prismicio\\";
import { components } from \\"@/slices\\";

type Params = { uid: string };

export default async function Page({ params }: { params: Params }) {
  const client = createClient();
  const page = await client
    .getByUID(\\"foo_bar\\", params.uid)
    .catch(() => notFound());

  return <SliceZone slices={page.data.slices} components={components} />;
}

export async function generateMetadata({
  params,
}: {
  params: Params;
}): Promise<Metadata> {
  const client = createClient();
  const page = await client
    .getByUID(\\"foo_bar\\", params.uid)
    .catch(() => notFound());

  return {
    title: page.data.meta_title,
    description: page.data.meta_description,
  };
}

export async function generateStaticParams() {
  const client = createClient();
  const pages = await client.getAllByType(\\"foo_bar\\");

  return pages.map((page) => {
    return { uid: page.uid };
  });
}
~~~

Make sure all of your import paths are correct. See the [install guide](https://prismic.io/docs/setup-nextjs) for more information.",
  "label": "App Router",
}
`;

exports[`PageSnippet > Pages Router > return a snippet with JavaScript 1`] = `
{
  "content": "## Create your Foo Bar's page component

Add a new route by creating a \`pages/[uid].js\` file. (If the route should be nested in a child directory, you can create the file in a directory, like \`pages/marketing/[uid].js\`.)

~~~js [pages/[uid].js]
import Head from \\"next/head\\";
import { isFilled, asLink } from \\"@prismicio/client\\";
import { SliceZone } from \\"@prismicio/react\\";

import { components } from \\"@/slices\\";
import { createClient } from \\"@/prismicio\\";

export default function Page({ page }) {
  return (
    <>
      <Head>
        <title>{page.data.meta_title}</title>
        {isFilled.keyText(page.data.meta_description) ? (
          <meta name=\\"description\\" content={page.data.meta_description} />
        ) : null}
      </Head>
      <SliceZone slices={page.data.slices} components={components} />
    </>
  );
}

export async function getStaticProps({ params, previewData }) {
  // The \`previewData\` parameter allows your app to preview
  // drafts from the Page Builder.
  const client = createClient({ previewData });

  const page = await client.getByUID(\\"foo_bar\\", params.uid);

  return {
    props: { page },
  };
}

export async function getStaticPaths() {
  const client = createClient();

  const pages = await client.getAllByType(\\"foo_bar\\");

  return {
    paths: pages.map((page) => {
      return asLink(page);
    }),
    fallback: false,
  };
}
~~~

Make sure all of your import paths are correct. See the [install guide](https://prismic.io/docs/setup-nextjs) for more information.",
  "label": "Pages Router",
}
`;

exports[`PageSnippet > Pages Router > return a snippet with TypeScript for TypeScript projects 1`] = `
{
  "content": "## Create your Foo Bar's page component

Add a new route by creating a \`pages/[uid].tsx\` file. (If the route should be nested in a child directory, you can create the file in a directory, like \`pages/marketing/[uid].tsx\`.)

~~~tsx [pages/[uid].tsx]
import { GetStaticPropsContext, InferGetStaticPropsType } from \\"next\\";
import Head from \\"next/head\\";
import { isFilled, asLink } from \\"@prismicio/client\\";
import { SliceZone } from \\"@prismicio/react\\";

import { components } from \\"@/slices\\";
import { createClient } from \\"@/prismicio\\";

type Params = { uid: string };

export default function Page({
  page,
}: InferGetStaticPropsType<typeof getStaticProps>) {
  return (
    <>
      <Head>
        <title>{page.data.meta_title}</title>
        {isFilled.keyText(page.data.meta_description) ? (
          <meta name=\\"description\\" content={page.data.meta_description} />
        ) : null}
      </Head>
      <SliceZone slices={page.data.slices} components={components} />
    </>
  );
}

export async function getStaticProps({
  params,
  previewData,
}: GetStaticPropsContext<Params>) {
  // The \`previewData\` parameter allows your app to preview
  // drafts from the Page Builder.
  const client = createClient({ previewData });

  const page = await client.getByUID(\\"foo_bar\\", params!.uid);

  return {
    props: { page },
  };
}

export async function getStaticPaths() {
  const client = createClient();

  const pages = await client.getAllByType(\\"foo_bar\\");

  return {
    paths: pages.map((page) => {
      return asLink(page);
    }),
    fallback: false,
  };
}
~~~

Make sure all of your import paths are correct. See the [install guide](https://prismic.io/docs/setup-nextjs) for more information.",
  "label": "Pages Router",
}
`;
